### PR TITLE
[Fix] make iou3d.boxes_iou3d actually calculate 3D IoU

### PR DESCRIPTION
**Pull request by filaPro**
## Motivation

Fix a bug introduced in #1902. Actually `boxes_iou3d` was calculating `bev iou`. Now it is fixed. Also `boxes_overlap_bev` is now a public function and can be used not only to calculate `3d iou` but also `3d iof` in `mmdetection3d`.

## Modification

Fix `boxes_iou3d` in python and cuda. Add `boxes_overlap_bev` and corresponding tests.

## BC-breaking (Optional)

No. Actually `boxes_iou3d` was not used in `mmdetection3d` and now it is just fixes.

May by @ZwwWayne or @Tai-Wang can have a look about `mmdetection3d` connection.

### CODE REVIEW

File: mmcv/ops/iou3d.py [hunk 1]
```diff
@@ -8,28 +8,62 @@
...
 ext_module = ext_loader.load_ext('_ext', [
-    'iou3d_boxes_iou3d_forward', 'iou3d_nms3d_forward',
+    'iou3d_boxes_overlap_bev_forward', 'iou3d_nms3d_forward',
     'iou3d_nms3d_normal_forward'
 ])
 
 
-def boxes_iou3d(boxes_a: Tensor, boxes_b: Tensor) -> Tensor:
+def boxes_overlap_bev(boxes_a: Tensor, boxes_b: Tensor) -> Tensor:
     """Calculate boxes 3D IoU.
```
Comment (Tai-Wang) [inline mmcv/ops/iou3d.py:17]: May adjust the docstring accordingly

**Comment by Tai-Wang** [comment]
Other modifications LGTM.

File: mmcv/ops/csrc/common/cuda/iou3d_cuda_kernel.cuh [hunk 2]
```diff
@@ -216,21 +216,21 @@ __device__ inline float iou_bev(const float *box_a, const float *box_b) {
...
-    const float *cur_box_a = boxes_a + a_idx * 7;
-    const float *cur_box_b = boxes_b + b_idx * 7;
-    float cur_iou_bev = iou_bev(cur_box_a, cur_box_b);
-    ans_iou[a_idx * num_b + b_idx] = cur_iou_bev;
+__global__ void iou3d_boxes_overlap_bev_forward_cuda_kernel(
+    const int num_a, const float *boxes_a, const int num_b,
+    const float *boxes_b, float *ans_overlap) {
+  // params boxes_a: (N, 7) [x, y, z, dx, dy, dz, heading]
+  // params boxes_b: (M, 7) [x, y, z, dx, dy, dz, heading]
+  const int a_idx = blockIdx.y * THREADS_PER_BLOCK + threadIdx.y;
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/iou3d_cuda_kernel.cuh:224]: The block you used to launch the kernel is:
```c++
  dim3 blocks(GET_BLOCKS(num_b, THREADS_PER_BLOCK_IOU3D),
              GET_BLOCKS(num_a, THREADS_PER_BLOCK_IOU3D));
```

`THREADS_PER_BLOCK` is not safe here.

**Review by Tai-Wang — APPROVED** [review]

**Review by grimoire — APPROVED** [review]
LGTM

**Pull request merged**